ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -176,7 +194,38 @@ Uploader::set_runtime_id(std::string_view id)
 }
 
 bool
+Uploader::thread_upload_impl(const Profile* profile)
+{
+    // As per convention, we wrap the thread so that all return paths
+    // mark the termination of the thread.
+    // There's a race condition here, because the thread will still be
+    // alive for several cycles after the return of this, but it shouldn't
+    // matter since that context is non-consuming and in a pre-join state.
+    bool ret = upload_impl(profile);
+    thread_working.store(false);
+    return ret;
+}
+
+void
 Uploader::upload(const Profile* profile)
+{
+    // If the upload thread is busy, give it some time.  We give it up to 3 seconds, which is arbitrary
+    wait_for_thread(3);
+    if (upload_thread && upload_thread->joinable()) {
+        upload_thread->join();
+        upload_thread.reset();
+    }
+
+    // Launch the thread.  We mark the working state here, potentially
+    // before the thread even launches, but the fence at the top of this
+    // function should ensure the thread has been created by the time
+    // joinability is checked
+    thread_working.store(true);
+    upload_thread.emplace(&Uploader::thread_upload_impl, this, profile);
```

Comment:
Apologies if I take a step back here, but I believe the upload is already happening in a periodic service (hence a periodic thread). Why do we have the need to start another thread to do the actual upload? Can't we just release the GIL while the upload here takes place? Pardon me if this has already been discussed, but I don't have any recollection of this at the moment 🙁

Reply:
The short answer:  it's not a need at all.  I didn't see a strong reason to go one way or the other, so I chose this one because I thought it would lead to a cleaner design.

The long answer:
There were a few minor reasons, but I confess that they're somewhat flimsy.

* I thought it would be easier to understand the behavior of the library ("ddup") as a unit if its thread dispatch and scheduling were internal.  I'm personally willing to accept a little bit more local complexity in order to be able to draw more rigid lines between components.
* Since the periodic scheduler is responsible for reacting to on_shutdown signals, I had a (very minor) concern about maintaining reactivity.
* I didn't see a particularly strong proscription against the dispatch of async operations within the periodic scheduler (this is such a one).  I also figured that from the kernel work-scheduling perspective, it was a wash either way (the periodic scheduler is hopefully sleeping when we're serializing+uploading).
* I'd like to cook up a more competent solution to transient HTTP errors (double-buffering is... not that solution), but I wanted to be able to think of the problem in terms that would fall under a single semantic umbrella (language/runtime).

WDYT?  What are the important factors at play here?

Reply:
My feeling is that if we went with the standard "GIL release on I/O" approach we could save the extra logic required to keep note of the thread activity. This is already done on the Python side by the PeriodicService class anyway. So I feel it would make the code slightly easier to reason about.

Reply:
All right.  It's an easy enough thing to move around that imho there's no harm in doing it that way.  I think I've gotten most of that fix landed, testing now.

Reply:
Oops, forgot to release the GIL.  Fixing.

Reply:
OK, should be better in 88828b4